src/System.Buffers.Primitives/System/Buffers/ReaderWriter/BufferReader.cs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Sequences;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public class BufferReader
+    {
+        public static BufferReader<TSequence> Create<TSequence>(TSequence buffer) where TSequence : ISequence<ReadOnlyMemory<byte>>
+        {
+            return new BufferReader<TSequence>(buffer);
+        }
+    }
+
+    public ref partial struct BufferReader<TSequence> where TSequence : ISequence<ReadOnlyMemory<byte>>
+    {
+        private TSequence _sequence;
+        private Position _currentSegmentPosition;
+        private Position _nextSegmentPosition;
+
+        private ReadOnlySpan<byte> _currentSegment;
+        private int _currentSegmentConsumedBytes;
+        
+        // TODO: this should be long, but it will impact the HTTP parser. We should meassure perf impact in isolation
+        private int _totalConsumedBytes;
```

Comment:
It has to be long to be correct, so it doesn't really matter if it does impact parser.

Reply:
will do